src/Algebra/Graph/Labelled/AdjacencyMap.hs

```diff
@@ -207,6 +207,23 @@ removeEdge x y = AM . Map.adjust (Map.delete y) x . adjacencyMap
 replaceVertex :: (Ord a, Semigroup e) => a -> a -> AdjacencyMap e a -> AdjacencyMap e a
 replaceVertex u v = gmap $ \w -> if w == u then v else w
 
+-- | Replace an edge from a given graph. If it doesn't exist, it will be created.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- replaceEdge e x y m                 == overlay (removeEdge x y m) (edge e x y)
+-- replaceEdge e2 x y (edge e1 x y)    == edge e2 x y
+-- edgeLabel x y (replaceEdge e x y m) == e
+-- @
+replaceEdge :: (Eq e, Monoid e, Ord a) => e -> a -> a -> AdjacencyMap e a -> AdjacencyMap e a
+replaceEdge e x y
+  | e == zero  = AM . createVertex . Map.alter (Just . maybe Map.empty (Map.delete y)) x . adjacencyMap
+  | otherwise  = AM . createVertex . Map.alter replace x . adjacencyMap
```

Comment:
I don't think you need `createVertex` here: 
* If `x` is not in the map, you'll add both `x` and `y` in the `replace Nothing` case.
* If `y` is not in the map, you'll add it in `Map.insert y e m`.

Reply:
Without `createVertex` it fails test on 
`replaceEdge e x y m == overlay (removeEdge x y m) (edge e x y)`
with input `Sum 1`, `0`, `1`, `empty`.

I think it creates edge from `x to y` in "inner" `Map` (successors of `x`) so `y` is there but there is no `y` in "outer" `Map` (all vertices)

Reply:
Ah, you are right! This illustrates how easy it is to make a mistake when you have invariants in your data structure. Whenever I'm working on `AdjacencyMap`s I have a feeling I'm walking a mine-field. Algebraic representation is so much safer :-)